### PR TITLE
grpc: make trace.FromContext work, and record handler status in traces

### PR DESCRIPTION
**Pull request by Sajmani**
Record the description of the status returned by server RPC
handlers in request traces, and mark the trace as an error if the
status is not OK.

Install the trace into the Context passed to server handlers using
trace.NewContext, so that code in the server handlers can annotate the
trace using trace.FromContext.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -396,10 +408,24 @@ func (s *Server) processStreamingRPC(t transport.ServerTransport, stream *transp
 			ss.statusDesc = appErr.Error()
 		}
 	}
+	if ss.tracing {
+		ss.mu.Lock()
+		if ss.statusCode != codes.OK {
+			ss.traceInfo.tr.LazyLog(stringer(ss.statusDesc), true)
+			ss.traceInfo.tr.SetError()
+		} else {
+			ss.traceInfo.tr.LazyLog(stringer("OK"), true)
```
Comment (dsymonds) [inline server.go]: This isn't sensitive info. This can be `ss.traceInfo.tr.LazyPrintf("OK")`. I wonder if some of your other uses of `stringer` are unnecessary too.

File: server.go [hunk 2]
```diff
@@ -396,10 +408,24 @@ func (s *Server) processStreamingRPC(t transport.ServerTransport, stream *transp
...
+		}
+		ss.mu.Unlock()
+	}
 	return t.WriteStatus(ss.s, ss.statusCode, ss.statusDesc)
 
 }
 
+type stringer string
```
Comment (dsymonds) [inline server.go]: Please put this in trace.go, next to `fmtStringer`. Or consider using that type directly instead.

File: stream.go [hunk 3]
```diff
@@ -283,7 +284,8 @@ type serverStream struct {
 	statusCode codes.Code
 	statusDesc string
 
-	tracing bool // set to EnableTracing when the serverStream is created.
+	tracing bool            // set to EnableTracing when the serverStream is created.
+	ctx     context.Context // provides trace.FromContext when tracing
```
Comment (dsymonds) [inline stream.go]: this grouping is confusing. at a glance it looks like this is only set when tracing, but that's not the case. I'd put this field higher up in this struct.

File: server.go (hunk 2 above)
Comment (Sajmani) [inline server.go]: Moved to trace.go

File: server.go (hunk 1 above)
Comment (Sajmani) [inline server.go]: Changed sensitive to false.  Not sure about switching to Printf, though; stringer avoids the printf and adds no struct overhead.  But really we should probably just have a way to add strings to a trace without requiring a stringer.

**Comment by Sajmani** [comment]
Updated with new commit.  Note that I've added another bug fix regarding context cancelation.

File: server.go (hunk 1 above)
Comment (dsymonds) [inline server.go]: LazyPrintf does exactly what LazyLog+fmtStringer does, except it doesn't have a control for sensitivity. There is indeed a struct overhead, though I suspect it would be negligible. Not too fussed.

File: server.go [hunk 4]
```diff
@@ -282,12 +282,15 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
 }
 
 func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc) (err error) {
+	ctx, cancel := context.WithCancel(stream.Context())
```
Comment (dsymonds) [inline server.go]: This seems an odd place to introduce this, no? Shouldn't whoever is creating that context be canceling it? Anything that gets to the `transport.Stream` will be getting a context differing from this one.

Comment (dsymonds) [inline server.go]: This seems like it should be addressed in its own change.

File: server.go [hunk 5]
```diff
@@ -365,9 +376,12 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
 }
 
 func (s *Server) processStreamingRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, sd *StreamDesc) (err error) {
+	ctx, cancel := context.WithCancel(stream.Context())
```
Comment (dsymonds) [inline server.go]: ditto

File: server.go (hunk 1 above)
Comment (Sajmani) [inline server.go]: We need stringer for LazyLog(stringer(statusDesc), true) so that we can set its sensitivity and ensure statusDesc is not interpreted as a format string.  While we could write LazyPrintf("OK"), I don't see why we should, given that LazyLog(stringer("OK"), false) does what we want.

But personally I'd prefer to have a Trace.Log(string) method to sidestep this whole issue.  I can send that in a separate CL if you're open to it.

File: server.go (hunk 4 above)
Comment (Sajmani) [inline server.go]: Interesting.  trasport.Stream has a cancel field to cancel its Context, but I cannot see when it's called.  Still hunting.

I need this change for my gRPC demo, but I can pull it out into a separate PR in parallel with this one.

Comment (Sajmani) [inline server.go]: *Stream.cancel is called from *http2Server.closeStream, which is called from a few different places, notably *http2Server.WriteStatus.  But WriteStatus returns early if the stream is already in state streamDone, which is set in 3 places.  I cannot tell whether those other code locations lead to calling *Stream.cancel, but it seems like they should.

Comment (Sajmani) [inline server.go]: David: I'd like to leave this PR as is but add a comment on these two context additions indicating that they should be removed when transport.Stream.Context is canceled properly.  I will file this as an issue against gRPC once I can pull together a repro case and link to that issue in the comments.  But I'd prefer to keep this PR as is so that the Context as seen by handlers is correct (and the demo works properly).

Comment (iamqizhao) [inline server.go]: @Sajmani SGTM given the timeline for your demo. You can leave this to me. I'll figure out a way to address it.

Comment (iamqizhao) [inline server.go]: I have slight concern that this introduces unnecessary latency on the fast path when tracing is disabled. We need a way to avoid it. But it is okay to put it into a separate pr.

File: server.go [hunk 6]
```diff
@@ -282,12 +282,15 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
...
 func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc) (err error) {
+	ctx, cancel := context.WithCancel(stream.Context())
+	defer cancel()
 	var traceInfo traceInfo
 	if EnableTracing {
 		traceInfo.tr = trace.New("grpc.Recv."+methodFamily(stream.Method()), stream.Method())
 		defer traceInfo.tr.Finish()
 		traceInfo.firstLine.client = false
 		traceInfo.tr.LazyLog(&traceInfo.firstLine, false)
+		ctx = trace.NewContext(ctx, traceInfo.tr)
```
Comment (iamqizhao) [inline server.go]: It seems if at https://github.com/grpc/grpc-go/blob/master/transport/http2_server.go#L181, you do

if EnableTracing {
  ctx = trace.NewContext(ctx, traceInfo.tr)  // create traceInfo there?
}

then tune line#183 and 185 a bit. All the concerns about cancellation are addressed.

Comment (Sajmani) [inline server.go]: How would server.go:handleStream get the traceInfo if it's created in http2_server.go?  Would you expose it via the *transport.Stream type?  I think this refactoring belongs in a separate PR.

File: server.go (hunk 4 above)
Comment (Sajmani) [inline server.go]: So can this remain unchanged?

File: server.go (hunk 5 above)
Comment (Sajmani) [inline server.go]: See discussion above; I think Qi Zhao has ideas on how to refactor.

**Comment by iamqizhao** [comment]
merged. I will try to refactor the code to address context cancellation issue by Wed.
